✨ Bonus Entries

Nico105 opened the pull request:
`console.log(userArray)` Giveaway.js Line 398 to see the array with the bonus entries

**Format**
```js
bonusEntries: [{ bonus: (member) => member.roles.cache.get('12345' ? 1 : null), cumulative: false }]
```

Thoughts, suggestions, improvements, roasts?
If someone has a better name for the property, I'm open.

Androz2091 commented:
I think we should use `serialize-javascript` to store the bonus entries in the JSON file and in the other databases.

```js
const serialize = require('serialize-javascript');

const bonusEntries = [
    {
        filter: (member) => member.roles.cache.some((r) => r.name === 'Booster'),
        bonus: 10
    }
]

const myStringBonusEntries = serialize(bonusEntries); // this is a simple string that can be stored anywhere

const retrievedBonusEntries = eval(`( ${myStringBonusEntries} )`);

// and retrievedBonusEntries is the same as bonusEntries!
```

I also added some commits to the PR, let me know if something was wrong

Nico105 commented:
> I think we should use `serialize-javascript` to store the bonus entries in the JSON file and in the other databases.
> 
> ```js
> const serialize = require('serialize-javascript');
> 
> const bonusEntries = [
>     {
>         filter: (member) => member.roles.cache.some((r) => r.name === 'Booster'),
>         bonus: 10
>     }
> ]
> 
> const myStringBonusEntries = serialize(bonusEntries); // this is a simple string that can be stored anywhere
> 
> const retrievedBonusEntries = eval(`( ${myStringBonusEntries} )`);
> 
> // and retrievedBonusEntries is the same as bonusEntries!
> ```
> 
> I also added some commits to the PR, let me know if something was wrong

Added and tested with mongoose and .json

Nico105 commented:
Or should the default manager bonusEntries and the giveawayObject bonus entries not be mergerd?
In the view of a abuser (=bot owner who gives himself 100 entries for every giveaway on every server his bot is in), it does not really matter because he can add it either way.
In terms of a honest bot owner it is more convient because you could say that for example if someone voted for your bot on top.gg then that member gets +1 entrie for every giveaway of the bot.
Or would that still be a bad thing? Idk

Lebyy commented:
When will this be implemented? Eagerly waiting for it

Nico105 commented:
> When will this be implemented? Eagerly waiting for it

Within this month (February) maybe?
Androz is a busy man

Lebyy commented:
> > When will this be implemented? Eagerly waiting for it
> 
> Within this month (February) maybe?
> Androz is a busy man

I can understand his life xD he has so many packages to maintain plus his bot(s).

Nico105 commented:
@Androz2091 I'm unsure, should we check giveaway object bonus entries AND `manager.default` bonus entries like for `exemptMembers`
Or should we check `manager.default` only when there are no giveaway object bonus entries like `exemptPermissions`?

https://github.com/Androz2091/discord-giveaways/pull/187/commits/01ad4009b271b40fdd3ec4b90543453c36733d73

Nico105 commented:
And I need an answer to my #169 review so that I know how we want to handle "custom arguments" because that would affect the `bonusEntries` format in the giveaway object.

Nico105 commented:
And should we maybe change the format to this: https://github.com/Androz2091/discord-giveaways/issues/108#issuecomment-699502945 ? 
Idk if it's more user friendly but then we wouldn't need bonus entry objects.

Nico105 commented:
And should we maybe add a `addable: true` option to the object?
Because currently we get all the bonus entry numbers from the functions and choose the highest of them as the amount of bonus Entries.
But with that extra property a dev would be able to say that if a member fulfills 2 functions then those entries should get added up `filter 1 = true, filter 2 = true | =>| 2 entries + 4 entries = 6 entries` `checkBonusEntries(), entries[6]`

Androz2091 commented:
Good idea, but I would use for `cumulative` instead of `addable`

Androz2091 commented on src/Giveaway.js:
```diff
@@ -308,6 +319,34 @@ class Giveaway extends EventEmitter {
+
+        if (this.bonusEntries.length) {
+            for (const obj of this.bonusEntries) {
+                const filter = obj[Object.keys(obj)[0]];
+                const bonus = obj[Object.keys(obj)[1]];
```
so here we should use properties instead of Object.keys

Nico105 replied:
Changed 👍

Androz2091 commented on src/Giveaway.js:
```diff
@@ -327,15 +366,44 @@ class Giveaway extends EventEmitter {
             else {
                 // Find a new winner
-                for (const user of users.array()) {
+                for (const user of userArray || user.array()) {
```
user is not defined

Nico105 replied:
oops

Androz2091 commented on typings/index.d.ts:
```diff
@@ -41,6 +41,10 @@ declare module 'discord-giveaways' {
         public emit<K extends keyof GiveawaysManagerEvents>(event: K, ...args: GiveawaysManagerEvents[K]): boolean;
     }
+    interface BonusEntry {
+        filter (member: GuildMember): boolean|Promise<boolean>;
+        bonus: number;
```
So here (and everywhere else for logic), it would become:
```suggestion
        bonus (member: GuildMember): number|Promise<number>;
        cumulative: boolean;
```

Androz2091 reviewed (approved):
Thank you for your awesome contribution to the project! :grin:
